Linear_Bin: function to bin spectrum by a non-integer number of pixels in each dimension.

k8macarthur opened the pull request:
Method to Bin EDS Spectrum by a non-integer number of pixels with linear interpolation and preservation of the original raw counts.

k8macarthur committed:
Merge remote-tracking branch 'hyperspy/master'

k8macarthur committed:
Merge remote-tracking branch 'hyperspy/master'

Linear_bin function added

k8macarthur committed:
Adding Linear_bin to eds.py

k8macarthur committed:
Adding Linear Bin Function for EDS

k8macarthur committed:
Merge pull request #3 from k8macarthur/Non_Integer_Binning

Non Integer Binning

k8macarthur committed:
Revert "Non Integer Binning"

k8macarthur committed:
Merge pull request #4 from k8macarthur/revert-3-Non_Integer_Binning

Revert "Non Integer Binning"

k8macarthur committed:
Merge branch 'master' into Non_Integer_Binning

k8macarthur committed:
Merge remote-tracking branch 'hyperspy/RELEASE_next_major' into Non_Integer_Binning

k8macarthur committed:
Revert "Revert "Non Integer Binning""

k8macarthur committed:
Merge pull request #5 from k8macarthur/revert-4-revert-3-Non_Integer_Binning

Revert "Revert "Non Integer Binning""

k8macarthur committed:
Merge branch 'master' into Non_Integer_Binning

k8macarthur commented:
Try again with changes actually included now...

k8macarthur commented:
So this is fully tested for EDS, would need to be copied across for EELS I guess but I put it in the folder with the other rebin function. I have included a function in the array section too but not tested. I also wasn't sure where the testing functions for this should go.

@thomasaarholt Happy to extend to 1D signals but maybe you could help with how to do that? Or just include your function?

thomasaarholt commented:
Does this work for Signal2Ds as well as Signal1Ds? EDS is a Signal1D due to its spectrum signal nature.

thomasaarholt commented:
And well  done!

thomasaarholt commented:
Linear_bin takes a "scale" argument, but in the doctext it describes a "step" argument. 

Also, could you include a small example in the doctext?

k8macarthur committed:
Fix Typo and Add Example to Docstring

k8macarthur commented:
Ok fixed the error in the doc_string and added an example.

In theory it could work Signal2Ds the function just extracts and plays with the np.array before putting it back into a Spectrum format (with amended meta data) and as there is no specification on the number of axis in the array. Although I haven't tested it yet. I think that there are so many different possible scenarios that this sort I function would be useful that I figured making a pull-request an allowing other people to test might be simpler than try to guess all the different scenarios myself. :)

thomasaarholt commented on hyperspy/_signals/eds.py:
```diff
@@ -232,6 +233,110 @@ def rebin(self, new_shape):
+        print(spectrum)
+        print ('Sum = ', sum(sum(sum(spectrum.data))))
+        scale = [2, 2, 5]
+        test = spectrum.linear_bin(step)
```
Your example includes the step typo ;)

k8macarthur replied:
Thanks :)

thomasaarholt commented on hyperspy/misc/array_tools.py:
```diff
@@ -96,11 +97,72 @@ def rebin(a, new_shape):
+    ----------
+    originalSpectrum: numpy.array
+        the original spectrum
+    step: a list of floats for each dimension specify the new:old pixel
```
Should be scale, here as well - copy the example from above, but as a function rather than a method?

k8macarthur replied:
Already spotted. :)

thomasaarholt reviewed (commented):
Found a few more things

k8macarthur commented:
So the function would need to copied and modified to work on 2D signals but I don't think it should be too much work.

thomasaarholt commented:
I might have a play with it at some point :)

We should write some tests for this. We could write a pretend signal and `linear_rebin` it to a known shape, then check the output. Any other smart things we could check?

k8macarthur commented:
The live_time should be checked as in:

`    ```
def test_rebin_live_time(self):
        s = self.signal
        old_metadata = s.metadata.deepcopy()
        dim = s.axes_manager.shape
        s = s.rebin([dim[0] / 2, dim[1] / 2, dim[2]])
        nt.assert_equal(
            s.metadata.Acquisition_instrument.TEM.Detector.EDS.live_time,
            3.1 \* 2 \* 2)
        # Check that metadata is unchanged
        print(old_metadata, self.signal.metadata)    # Captured on error
        nt.assert_dict_equal(old_metadata.as_dictionary(),
                             self.signal.metadata.as_dictionary(),
                             "Source metadata changed")

```
```

k8macarthur commented:
I have some basic tests already written just wasn't sure which file they should be going in...

k8macarthur committed:
Add set of test functions + typos

k8macarthur committed:
Documentation Added to the User Guide

thomasaarholt commented:
Getting this warning when binning by an non-integer:

```
/Users/thomas/Dropbox/0_Git/GitHub/hyperspy_aar/hyperspy/_signals/eds.py:315: VisibleDeprecationWarning: using a non-integer number instead of an integer will result in an error in the future
  newSpectrum[j] = newSpectrum[j] + s[bottomPos]
```

thomasaarholt commented:
If I try to bin a line profile I get an error:

``` python
>>> s
<EDSTEMSpectrum, title: EELS, dimensions: (70, 78|1995)>
>>> s2 = s.inav[0,:]
<EDSTEMSpectrum, title: EELS, dimensions: (78|1995)>
In [20]:
>>> s2.linear_bin([2,2])
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-22-1e64d6c62abd> in <module>()
----> 1 s2.linear_bin([2,2])

/Users/thomas/Dropbox/0_Git/GitHub/hyperspy_aar/hyperspy/_signals/eds.py in linear_bin(self, scale)
    295             s[:] = newSpectrum
    296             newSpectrum = np.zeros((math.ceil(shape2[0]/step),
--> 297                                     shape2[1], shape2[2]), dtype='float')
    298             if k != 0:
    299                 s = np.swapaxes(s, 0, k)

IndexError: tuple index out of range
```

thomasaarholt commented:
I'm about to submit a pull request which should enable support for line profiles and Signal2Ds. I'm also moving the method to the BaseSignal class, so that it is available to all types of signals.
I would be submitting it in the next ten minutes, but my laptop just ran out of power. Charging it tonight.

thomasaarholt commented:
Added a pull request to your repository. :)

https://github.com/k8macarthur/hyperspy/pull/6

k8macarthur committed:
Limiting main linear_bin function to one location

k8macarthur committed:
Correction to Live_time/real_time adjustment to only spatial dimension binning

thomasaarholt commented:
Huh. Neat. That was a nice idea.

thomasaarholt commented on hyperspy/misc/array_tools.py:
```diff
@@ -123,18 +123,19 @@ def linear_bin(s, scale):
+    newSpectrum = s[:]
 
-    if len(shape) != len(scale):
+    if len(shape2) != len(scale):
```
The line `newSpectrum = s[:]` entirely overwrites the variable `newSpectrum`, meaning that the previous line doesn't do anything at all. If you wanted to fill `newSpectrum` with the values in the `s` array you would do:
`newSpectrum[:] = s[:]`, but I'm not sure that's necessary at all.

Shouldn't `newSpectrum = s` be enough?

thomasaarholt commented:
If it proves difficult, then I'll just keep a copy of my code in a second document and just update your modified array_tools function.

thomasaarholt reviewed (commented):
Just a small note

k8macarthur commented:
In theory it should be enough but then I kept getting complaints that I was trying to perform all the following functions on a spectrum rather than an array...

> On 13 Oct 2016, at 17:45, Thomas Aarholt notifications@github.com wrote:
> 
> @thomasaarholt commented on this pull request.
> 
> Just a small note
> 
> In hyperspy/misc/array_tools.py https://github.com/hyperspy/hyperspy/pull/1305#pullrequestreview-4102884:
> 
> > -    if len(shape) != len(scale):
> > -    if len(shape2) != len(scale):
> >   Line 127: The line newSpectrum = s[:] entirely overwrites the variable newSpectrum, meaning that the previous line doesn't do anything at all. If you wanted to fill newSpectrum with the values in the s array you would do:
> >   newSpectrum[:] = s[:], but I'm not sure that's necessary at all.
> 
> Shouldn't newSpectrum = s be enough?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub https://github.com/hyperspy/hyperspy/pull/1305#pullrequestreview-4102884, or mute the thread https://github.com/notifications/unsubscribe-auth/APMByfatg1KjPn8_N4r-LXOz0Hd-uQJVks5qzlItgaJpZM4KUitn.

k8macarthur committed:
Merge Change from Tom

k8macarthur commented:
@thomasaarholt if you're happy for the new structure do you want to add the new function for EELS?

thomasaarholt commented:
Yep, good idea. I'll learn from the way you added it and copy it over.

k8macarthur committed:
Remove function from signal.py

thomasaarholt commented:
Another thought: Should `scale` be a tuple instead of a list? It typically is, if you look at any type of `shape`.

k8macarthur committed:
Typo Fixing

thomasaarholt committed:
Added overwriteable-method to BaseSignal

I’m going to ask the rest of the Hyperspy group if this is bad
practice, but I think it should be fine and certainly works alright. If
a signal has type EDS, the method in the class overwrites the
BaseSignal one. So this way all signals have access.

thomasaarholt committed:
Added EELS method

thomasaarholt committed:
Added scaling of EELS dwell time

thomasaarholt committed:
Added EELS import of _linear_bin

thomasaarholt committed:
Reverse ordering of signal_dimension binning factors

This is due to hyperspy’s ordering of signal data. To check, try the
following:

```python
import hyperspy.api as hs
%matplotlib inline

import scipy
img = hs.signals.Signal2D(scipy.misc.face()).as_signal2D((2,0))
img.plot()
img.linear_bin([1,1,2]).plot()
```

thomasaarholt committed:
scale should be multiplied, not divided, by binning factor.

thomasaarholt committed:
Revert "scale should be multiplied, not divided, by binning factor."

This reverts commit e269a556004344a2327751ef29cab02d4210b338.

thomasaarholt committed:
Scale should be multiplied, not divided.

Fixed for EDS, EELS and BaseSignal
```python
s2 = s.linear_bin([20, 5, 1])

for axes in s.axes_manager._axes:
    print(axes.size*axes.scale + axes.offset)
for axes in s2.axes_manager._axes:
    print(axes.size*axes.scale + axes.offset)
```

thomasaarholt committed:
[optional] Removing unnecessary code

So, I run the code fine without these two lines.

thomasaarholt committed:
Same as previous commit for eels and eds

k8macarthur committed:
Merge pull request #7 from thomasaarholt/linbin3

Non-Integer Binning 2nd Amendment

k8macarthur committed:
Add option for dwell_time or exposure correction in EELS.

k8macarthur commented:
Hi @thomasaarholt , I checked the EELS metadata, and it looks like EELS can have dwell_time or exposure so I've updated to two if statements.

k8macarthur committed:
Update to description of return.

k8macarthur commented:
Also scale can be a tuple if you prefer. I have no preference.

k8macarthur commented:
So I think looking back at #1170 I should close this pull request and resubmit to release next minor. I'm going to try this now.

k8macarthur commented:
#1306
